core/src/main/java/org/elasticsearch/search/aggregations/metrics/percentiles/ParsedPercentiles.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.search.aggregations.metrics.percentiles;
+
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.ParsedAggregation;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+public abstract class ParsedPercentiles extends ParsedAggregation implements Percentiles {
+
+    private final Map<Double, Double> percentiles = new HashMap<>();
+    private final Map<Double, String> percentilesAsString = new HashMap<>();
+
+    protected boolean keyed;
+
+    void addPercentile(double key, double value) {
+        percentiles.put(key, value);
+    }
+
+    void addPercentileAsString(double key, String valueAsString) {
+        percentilesAsString.put(key, valueAsString);
+    }
+
+    @Override
+    public double percentile(double percent) {
+        Double percentile = percentiles.get(percent);
+        if (percentile == null) {
+            return Double.NaN;
+        }
+        return percentile;
+    }
+
+    @Override
+    public String percentileAsString(double percent) {
+        return percentilesAsString.get(percent);
+    }
+
+    @Override
+    public Iterator<Percentile> iterator() {
+        return new Iterator<Percentile>() {
+            final Iterator<Map.Entry<Double, Double>> iterator = percentiles.entrySet().iterator();
+            @Override
+            public boolean hasNext() {
+                return iterator.hasNext();
+            }
+
+            @Override
+            public Percentile next() {
+                Map.Entry<Double, Double> next = iterator.next();
+                return new InternalPercentile(next.getKey(), next.getValue());
```

Comment:
I also thought about it when I wrote this. I think it makes sense to use InternalPercentile here and I'm in favor of merging the implementation with its interface in core.